docs: Add how-to guide on separating cocotb and simulator logs (#5701)
cocotb/cocotb: docs/source/separating_logs.rst

.. _separating-logs:

******************************************
Separating cocotb and Simulator Log Output
******************************************

By default, cocotb's log output and the simulator's own output are interleaved
on the same stream.
This happens because the two arrive there by different routes:

* cocotb emits its messages through the Python :mod:`logging` module.
  :func:`cocotb.logging.default_config` calls :func:`logging.basicConfig` with
  ``stream=sys.stdout``, so records propagate to the root logger and are
  written to standard output.
* The simulator writes its own output, such as that produced by ``$display``,
  directly to standard output. It never passes through Python's
  :mod:`logging` machinery, so it cannot be redirected by configuring a logger.

> To separate the two, attach a handler to cocotb's own loggers and stop those

Reviewer: ```suggestion
To separate the two, attach a :class:`python:logging.Handler` to cocotb's own loggers and stop those
```

Author: This line no longer exists — the rewrite drops the `propagate = False` approach in favor of replacing the root logger's handlers, so the sentence it applied to is gone. I did adopt the `:class:`python:logging.Handler`` intersphinx form from your suggestion and used it throughout the page.

.. _separating-logs:

******************************************
Separating cocotb and Simulator Log Output
******************************************

When a test runs, cocotb's own log messages and the simulator's output
share a single stream.
This is fine interactively,
but it makes life hard when you want to keep one and not the other:
saving cocotb's log to a file for a CI system to parse,

> or keeping the simulator transcript free of cocotb chatter.

Reviewer: ```suggestion
but it makes life hard when you want to keep the simulator transcript free of cocotb chatter or vice versa.
```

We do *not* endorse parsing the log outputs, it is not a stable API.

This guide is for users who want cocotb's log output in a separate place
from the simulator's own output.
It explains why the two are mixed together,

> and how to send cocotb's log elsewhere.

Reviewer: Be concrete about what the guide is showing.
```suggestion
This guide shows users how to configure cocotb's loggers to output to a file.
```

Why the two are mixed
=====================

The two streams arrive at the same place by different routes.

> 

Reviewer: Redundant.
```suggestion
```

cocotb emits its messages through the Python :mod:`python:logging` module.
:func:`cocotb.logging.default_config` configures the root logger to write to
standard output,
and every cocotb logger propagates its records up to that root logger.

The simulator writes its own output, such as that produced by ``$display``,
directly to standard output.
It never passes through Python's :mod:`python:logging` machinery,

> so it cannot be redirected by configuring a logger.

Reviewer: Good addition, staves off follow-on questions.

Redirecting cocotb's log
========================

Because every cocotb logger propagates to the root logger,
you do not need to touch the individual cocotb loggers.
Remove the handlers cocotb installed on the root logger,
and register your own instead.

> cocotb's records then follow your handler,

Reviewer: ```suggestion
cocotb's logs then flow through your handler to wherever that handler is configured to write to,
```

while the simulator's output is left on standard output untouched.

.. code-block:: python

    import logging

    from cocotb.logging import SimLogFormatter

    root_logger = logging.getLogger()
    for handler in list(root_logger.handlers):
        root_logger.removeHandler(handler)
        handler.close()

    file_handler = logging.FileHandler("cocotb.log", mode="w")
    file_handler.setFormatter(SimLogFormatter())
    root_logger.addHandler(file_handler)

Place this at module level in your test module,

> so that it runs when cocotb imports the module.

Reviewer: Unneeded comma.
```suggestion
Place this at module level in your test module so that it runs when cocotb imports the module.
```

Using :class:`~cocotb.logging.SimLogFormatter` keeps the simulation timestamp
and the familiar cocotb layout in the redirected output.
A plain :class:`python:logging.Formatter` works too,
but the simulation time is then lost unless the format string refers to it.

To send cocotb's output somewhere other than a file,
register a different :class:`python:logging.Handler`;

> only the handler construction changes.

Reviewer: Redundant clause.
```suggestion
To send cocotb's output somewhere other than a file,
register a different :class:`python:logging.Handler`.
```

For example, a :class:`python:logging.StreamHandler` on :data:`python:sys.stderr`
keeps cocotb's messages on the console
while still separating them from the simulator's output on standard output,
which is convenient when redirecting one stream to a file from the shell.

.. note::

>    A handful of start-up messages are still written to standard output.

Reviewer: Thanks for the note, this preempts questions about why these are missing.

   These come from the ``gpi`` and ``pygpi`` loggers,
   and from ``cocotb.initialize`` reporting the cocotb and simulator versions.
   They are emitted before cocotb imports your test module,
   so they are logged before the handler above is installed.
   The exact count depends on the simulator,
   so treat the remaining start-up lines as expected rather than fixed.

> This recipe has been verified with Icarus Verilog and with Cadence Xcelium.

Reviewer: Unnecessary as this works through Python so it is always supported, regardless of the simulator.
```suggestion
```

In both cases the simulator's own ``$display`` output stayed on standard output

> while cocotb's records went to the file.

Reviewer: Let's just remove this whole paragraph.
```suggestion
```
